sdk/storage/Azure.Storage.DataMovement/src/DataTransferStatus.cs

```diff
@@ -1,66 +1,147 @@
 
         /// <summary>
-        /// The Job has been paused. When transfer is paused (e.g. see <see cref="TransferManager.PauseTransferIfRunningAsync(string, System.Threading.CancellationToken)"/>) during the transfer,
-        /// this will be the value.
+        /// Represents if transfer has any skipped items.
         ///
-        /// This status is a resumable state, only
-        /// transfers that failed will be retried when <see cref="TransferManager.StartTransferAsync(StorageResource, StorageResource, DataTransferOptions, CancellationToken)"/>
-        /// with the respective transfer ID to resume.
+        /// If set to `true`, the transfer has at least one item it has skipped.
+        /// If set to `false`, the transfer currently has no items that has been skipped.
+        ///
+        /// It's possible to never have any items skipped if
+        /// <see cref="StorageResourceCreationPreference.SkipIfExists"/> is not enabled in the <see cref="DataTransferOptions.CreationPreference"/>.
         /// </summary>
-        Paused = 3,
+        public bool HasSkippedItems { get; internal set; }
 
         /// <summary>
-        /// The Job has completed successfully with no failures or skips.
+        /// Constructor to set the initial state to <see cref="DataTransferState.Queued"/> with no failures or skipped items.
         /// </summary>
-        Completed = 4,
+        protected internal DataTransferStatus()
+        {
+            State = DataTransferState.Queued;
+            HasFailedItems = false;
+            HasSkippedItems = false;
+        }
 
         /// <summary>
-        /// The Job has been completed with at least one skipped transfer.
+        /// Constructor to have a custom state, failure state, and skipped state.
         /// </summary>
-        CompletedWithSkippedTransfers = 5,
+        protected internal DataTransferStatus(DataTransferState state, bool hasFailureItems, bool hasSkippedItems)
+        {
+            State = state;
+            HasFailedItems = hasFailureItems;
+            HasSkippedItems = hasSkippedItems;
+        }
+
+        internal bool IsCompletedWithFailedItems => State.Equals(DataTransferState.Completed) && HasFailedItems;
+        internal bool IsCompletedWithSkippedItems => State.Equals(DataTransferState.Completed) && HasSkippedItems;
 
         /// <summary>
-        /// The Job has been completed with at least one failed transfer.
+        /// Accordingly update the <see cref="HasFailedItems"/> to true. If already set to true, nothing will happen.
+        ///
+        /// This should only be triggered when a failed item has been seen.
+        /// </summary>
+        /// <returns>True if <see cref="HasFailedItems"/> was updated. False otherwise.</returns>
+        internal bool TrySetFailedItem()
+        {
+            if (!HasFailedItems)
+            {
+                HasFailedItems = true;
+                return true;
+            }
+            return false;
+        }
```

Comment:
I think I'm still a little confused as to what "try" is accomplishing here.

Also, I see elsewhere you're basically using this to determine if this is the first failed event such that you can fire a onetime event. But you'll be in a race condition for that. While the race condition doesn't matter for setting the value, it does for determining whether you changed the value. You'll need Interlocked for that, where you can atomically set the value and get back the prior value that was there.

Reply:
It "tries" to set it, if it hasn't been set already.

Wait. I actually had a lock here but you said it was unnecessary cause we didn't care about the race condition. So do we need a lock here or not?

Reply:
I'm just going to add the lock back. I definitely had it here, but I removed it cause of the feedback I got.

Reply:
Ah okay I see your point, we have to use `Interlocked.Exchange` to get back the original value. Then we can just return that. I will also apply it to the rest.